File: __tests__/features/bills/billReviewScreenData.test.ts
import { projectBillReview } from '@/features/bills/ui/billReviewScreenData';
import { asyncData, asyncError, asyncLoading } from '@/core/async/async';
import { makeCatalog } from '@/features/billers/data/billerCatalog';
import { fixedClock } from '@/core/clock/clock';

const catalog = makeCatalog([{ id: 'electricity', name: 'Electricity' }, { id: 'dth', name: 'DTH' }], [
  { id: 'electricity-metro', categoryId: 'electricity', name: 'Metro Electricity', mode: 'presentment', inputParams: [] },
  { id: 'dth-metro', categoryId: 'dth', name: 'Metro DTH', mode: 'openAmount', inputParams: [] },
]);
const clock = fixedClock(new Date(2026, 5, 5, 10));
const bill = { billerId: 'electricity-metro', account: 'K123', customerName: 'Ramesh Kumar', amountPaise: 45000, dueDate: new Date(2026, 5, 8), billNumber: 'BILL-X-1' };

describe('projectBillReview', () => {
  it('catalog loading is loading', () => {
    expect(projectBillReview(asyncLoading(), undefined, { billerId: 'electricity-metro', account: 'K123', amountPaise: null }, false, clock).kind).toBe('loading');
  });
  it('catalog error is error', () => {
    expect(projectBillReview(asyncError(new Error('x')), undefined, { billerId: 'electricity-metro', account: 'K123', amountPaise: null }, false, clock).kind).toBe('error');
  });
  it('openAmount uses the passed amount without a fetched bill', () => {
    const r = projectBillReview(asyncData(catalog), undefined, { billerId: 'dth-metro', account: 'D77', amountPaise: 25000 }, false, clock);
    expect(r.kind === 'review' && r.amountPaise).toBe(25000);
    expect(r.kind === 'review' && r.customerName).toBeNull();
    expect(r.kind === 'review' && r.dueInDays).toBeNull();
    expect(r.kind === 'review' && r.canPay).toBe(true);
  });
  it('openAmount with missing amount is an error', () => {
    expect(projectBillReview(asyncData(catalog), undefined, { billerId: 'dth-metro', account: 'D77', amountPaise: null }, false, clock).kind).toBe('error');
  });
  it('presentment review uses the fetched bill and computes dueInDays', () => {
    const r = projectBillReview(asyncData(catalog), asyncData(bill), { billerId: 'electricity-metro', account: 'K123', amountPaise: null }, false, clock);
    expect(r.kind === 'review' && r.customerName).toBe('Ramesh Kumar');
    expect(r.kind === 'review' && r.dueInDays).toBe(3);
    expect(r.kind === 'review' && r.amountPaise).toBe(45000);
  });
  it('paying disables pay', () => {
    const r = projectBillReview(asyncData(catalog), undefined, { billerId: 'dth-metro', account: 'D77', amountPaise: 25000 }, true, clock);
    expect(r.kind === 'review' && r.paying).toBe(true);
    expect(r.kind === 'review' && r.canPay).toBe(false);
  });
});

File: test/features/bills/bill_review_screen_data_test.dart
import 'package:billpayments/core/mock/mock_network.dart';
import 'package:billpayments/features/auth/data/auth_provider.dart';
import 'package:billpayments/features/billers/data/biller.dart';
import 'package:billpayments/features/billers/data/biller_catalog.dart';
import 'package:billpayments/features/billers/data/biller_catalog_provider.dart';
import 'package:billpayments/features/billers/data/biller_repository.dart';
import 'package:billpayments/features/bills/data/fetched_bill_provider.dart';
import 'package:billpayments/features/bills/ui/bill_review_screen_data.dart';
import 'package:billpayments/features/bills/ui/bill_review_screen_data_provider.dart';
import 'package:billpayments/features/payments/data/payments_provider.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';

class _ExplodingBillerRepository implements BillerRepository {
  @override
  Future<BillerCatalog> fetchCatalog(String language) async =>
      throw Exception('server down');

  @override
  Future<List<Biller>> search(String query, String language) async =>
      const [];
}

ProviderContainer instantContainer() {
  final container = ProviderContainer(overrides: [
    mockNetworkProvider
        .overrideWithValue(MockNetwork(minDelayMs: 0, maxDelayMs: 1)),
  ]);
  addTearDown(container.dispose);
  return container;
}

Future<void> login(ProviderContainer container) async {
  final notifier = container.read(authProvider.notifier);
  await notifier.sendOtp('9876543210');
  await notifier.verifyOtp('123456');
}

void main() {
  test('presentment review loads fetched bill with formatted amount',
      () async {
    final container = instantContainer();
    await login(container);
    const params = (
      billerId: 'electricity-metro',
      account: 'K123',
      amountPaise: null
    );
    final sub =
        container.listen(billReviewScreenDataProvider(params), (_, _) {});
    expect(container.read(billReviewScreenDataProvider(params)),
        const BillReviewScreenData.loading());
    await container.read(billerCatalogProvider.future);
    await container.read(
        fetchedBillProvider((billerId: 'electricity-metro', account: 'K123'))
            .future);
    await container.pump();
    final review = container.read(billReviewScreenDataProvider(params))
        as BillReviewLoaded;
    expect(review.billerName, 'Metro Electricity');
    expect(review.customerName, isNotNull);
    expect(review.dueInDays, isNotNull);
    expect(review.amountPaise, isA<int>());
    expect(review.canPay, isTrue);
    expect(review.paying, isFalse);
    sub.close();
  });

  test('openAmount review uses the passed amount without fetching', () async {
    final container = instantContainer();
    await login(container);
    const params = (billerId: 'dth-metro', account: 'D77', amountPaise: 25000);
    final sub =
        container.listen(billReviewScreenDataProvider(params), (_, _) {});
    await container.read(billerCatalogProvider.future);
    await container.pump();
    final review = container.read(billReviewScreenDataProvider(params))
        as BillReviewLoaded;
    expect(review.amountPaise, 25000);
    expect(review.customerName, isNull);
    expect(review.dueInDays, isNull);
    sub.close();
  });

  test('catalog failure surfaces as review error, not endless loading',
      () async {
    final container = ProviderContainer(
      retry: (_, _) => null,
      overrides: [
        mockNetworkProvider
            .overrideWithValue(MockNetwork(minDelayMs: 0, maxDelayMs: 1)),
        billerRepositoryProvider
            .overrideWithValue(_ExplodingBillerRepository()),
      ],
    );
    addTearDown(container.dispose);
    await login(container);
    const params = (
      billerId: 'electricity-metro',
      account: 'K123',
      amountPaise: null
    );
    final sub =
        container.listen(billReviewScreenDataProvider(params), (_, _) {});
    await expectLater(
        container.read(billerCatalogProvider.future), throwsException);
    await container.pump();
    expect(container.read(billReviewScreenDataProvider(params)),
        isA<BillReviewError>());
    sub.close();
  });

  test('review shows paying while a payment for this bill is processing',
      () async {
    final container = instantContainer();
    await login(container);
    const params = (billerId: 'dth-metro', account: 'D77', amountPaise: 25000);
    final sub =
        container.listen(billReviewScreenDataProvider(params), (_, _) {});
    await container.read(billerCatalogProvider.future);
    await container.pump();
    final paymentId = container.read(paymentsProvider.notifier).pay(
        billerId: 'dth-metro',
        billerName: 'Metro DTH',
        categoryId: 'dth',
        account: 'D77',
        amountPaise: 25000);
    expect(paymentId, isNotNull);
    final during = container.read(billReviewScreenDataProvider(params))
        as BillReviewLoaded;
    expect(during.paying, isTrue);
    expect(during.canPay, isFalse);
    await Future<void>.delayed(const Duration(milliseconds: 25));
    sub.close();
  });
}
